### ISSUE TITLE
Security issue with NSEC in the session storage. The NSEC is very easy to find and its a risk! 

### ISSUE THREAD

**Issue by julianomarx** [SnowCait/nostter#1434]
Hi guys, how you doing? 

I found a issue on the nostter.app client.... 

When we use the dev-tools of the browser, we can easily find the NSEC in the sessionStorage... its just a raw NSEC, without being encrypted o locked with a password... 

I think its a problem, beacuse its very very easy to access the sessionstorage... someone can take advantage of this problem to take the NSEC of the users... 

I did the same test in the SNORT.SOCIAL, and I cant find...  I dont know how the snort social did to keep this NSEC secure, but I think we need to review this.

![Captura de tela 2024-09-29 010622](https://github.com/user-attachments/assets/5a65fd36-9709-4b00-8506-c160e7e957e9)

Thanks for your time. 

Best regards!

**Comment by SnowCait** [comment]
Hi,

I also don't think it's a good idea to store private keys in local storage, but there's no way for others to access local storage except through extensions.
And if you use extensions, you can use NIP-07.
Please let me know if there are any other ways to access it.

Thanks.

**Comment by julianomarx** [comment]
Hi! 

I'll try to figure out how to encrypt these nsec's on the sessionstorage. I know that there's a way to do it, but I still don't know how to do it. 
It seems that we need to use a "password/passphrase" to decrypt the nsec just in the moment that we are going to sign an event. 
I'll try to understand how to do it and I'll tell you latter. 

Have a good day!

By the way... se you there, follow me if you feel confortable to. 

npub1du57w98d4sgjzeqz8wntk7w09855txtk87m45s4r9uy7xeare96qrat5sk

**Comment by SnowCait** [comment]
Before considering using a password, could you please prove that there is a way for others to access the local storage?
I'm not a frontend expert, so I don't know if it's really possible to access local storage from extensions.
If there isn't a way, it is meaningless.

I don't want to use a password if possible, as it would ruin the simplicity of Nostr.

**Comment by julianomarx** [comment]
Sure, no problem! 

I'll keep in touch.

**Comment by VnUgE** [comment]
> I'm not a frontend expert, so I don't know if it's really possible to access local storage from extensions.
If there isn't a way, it is meaningless.

Web extension content scripts run in the same window context. They have access to the page's local storage and session storage tmk

**Comment by SnowCait** [comment]
@VnUgE 
Can the extension access local storage directly? I'd like to know the specific code.
If not, it seems pointless to encrypt nsec with a password, since it can't be completely hidden as you say.

**Comment by VnUgE** [comment]
Content scripts seem to have minimal access to the window and dom they are running in (usually all windows). However, content scripts can inject other scripts (blobs) into the running window and marshal data using the messaging api. This is how nip07 works. 

My Example: 
https://github.com/VnUgE/NVault/blob/develop/extension/src/entries/contentScript/util.ts

This has the power to inject local scripts (as blobs) into all running doms which will have access to local and session storage for any site that is loaded. That script is designed to modify the window object, which can see `window.localStorage` and `window.sessionStorage`. The data still must be marshaled out of the window, 

I was fairly confident that both content scripts and background scripts also had this power, through other APIs, but I cannot find the API among mozilla docs. 

So that said, I think it's fair to assume an extension could always have the power to access all or most attributes of running windows if it wants. I'd argue that's the point of client side extensions. But I also don't think it's a web-app's responsibility to guard against (work against) malicious web extensions either. 

[Docs for content scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts#permissions_restrictions_and_limitations)

**Comment by SnowCait** [comment]
@julianomarx @VnUgE 
After some research, I confirmed that it is possible to access local storage from the content scripts of the extension.
However, as mentioned above, since nsec is stored in memory from local storage or other sources (even if it is protected by a password), it is impossible to completely hide it, and using password is not a fundamental solution.

I also checked other web clients, and they also store nsec in local storage. (Including snort.social, which has the raw hex in the `sessions` key in local storage.)
Some clients offer a password option.
There seems to be no way to solve this issue other than prohibiting login with private key.
